daemon/protocol/client_lib.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/env python
+
+import os
+import re
+import tempfile
+import shutil
+
+from grpc.beta import implementations
+#import grpc
+import daemon_pb2
+
+
+# Copied from daemon/protocol/locate.go
+def _locate_scoot_dir():
+    scootdir = os.environ.get("SCOOTDIR")
+    if scootdir is not None:
+        return scootdir
+    homedir = os.environ.get("HOME")
+    if homedir is not None:
+        return os.path.join(homedir, ".scoot")
+    return None
+
+def _locate_socket():
+    scootdir = _locate_scoot_dir()
+    if scootdir is None:
+        return None
+    return os.path.join(scootdir, "socket")
+
+
+# Singleton variables.
+_domain_sock = _locate_socket()
+_client = None
+_context = None
+
+class ScootException(Exception):
+    """
+    """
+    def __init__(self, caught):
+        self.caught = caught
+
+
+
+class ScootStatus:
+    """
+    """
+    class State:
+        UNKNOWN = 0
+        PENDING = 1
+        PREPARING = 2
+        RUNNING = 3
+        COMPLETED = 4
+        FAILED = 5
+    def __init__(self, run_id, state, snapshot_id, exit_code, error):
+        self.run_id = run_id
+        self.state = state
+        self.snapshot_id = snapshot_id
+        self.exit_code = exit_code
+        self.error = error
+
+
+def start():
+    """
+    """
+    global _client, _domain_sock
+    if _domain_sock is None:
+        raise ScootExceptoin("Could not find domain socket path.")
+    if is_started():
+        raise ScootException("Already started.")
+    #channel = grpc.insecure_channel("unix://%s:0" % _domain_sock)
+    #_client = daemon_pb2.ScootDaemonStub(channel)
+    channel = implementations.insecure_channel(host=_domain_sock, port=None)
+    _client = daemon_pb2.beta_create_ScootDaemon_stub(channel)
+
+
+
+def stop():
+    """
+    """
+    global _client
+    if not is_started():
+        raise ScootException("Already stopped.")
+    _client = None
+    pass
+
+
+
+def is_started():
+    """
+    """
+    global _client
+    return _client is not None
+
+
+def echo(ping):
+    """
+    """
+    global _client
+    if not is_started():
+        raise ScootException("Not started.")
+    req = daemon_pb2.EchoRequest(ping=ping)
+    try: resp = _client.Echo(req)
+    except Exception as e: raise ScootException(e)
+    return resp.pong
+
+
+def create_snapshot(path):
+    """
+    """
+    global _client, _context
+    if not is_started():
+        raise ScootException("Not started.")
+    req = daemon_pb2.CreateSnapshotRequest(path=path)
+    try: resp = _client.CreateSnapshot(req, _context)
+    except Exception as e: raise #ScootException(e)
```

Comment:
seems like this should also wrap with `ScootException`? maybe this was commented out for debugging?